### PR TITLE
Added method return chaining, custom matchers and a helper class “Is”.

### PR DESCRIPTION
**Pull request by cwilkinsonffdc**

### CODE REVIEW

**Comment by afawcett** [comment]
Wow quite a bit of functionality here, thanks! :+1: 

Would you mind adding a summary comment and updating your title here to describe what you've added support for ('is' support?) and some background etc. 

@comic96 is on holiday at present, but i'll for sure give him a heads up when he returns to review this excellent submission! 

Thanks again!

**Comment by cwilkinsonffdc** [comment]
@afawcett – thanks for looking at this for me, much appreciated.  :)

I’d say this pull request has 3 main areas of change. 

1)	Method return chaining. A simple change that allows us apex peeps to call a mocked method a number of times and get different returns. 

2)	Custom matchers. Currently you mock a method by saying, when method X gets called with param Y return Z. With custom matchers you can say, when method X is called, check the param and see if it matches some custom logic. If true, return Z. Also, these matchers preserve the param you called with. Very handy for asserting that an inner private method say, called you mock with the param “Zebra” or you can get the SObject passed and assert is had certain properties set. I know you can work around this by having well structured code. But for brown field development, this is very useful!

3)	A helper class “Is” to provide some pre built matchers. For example, Is.AnyInteger() or Is.AnyThing(). Sometimes you don’t care what got passed in. You just need to know it was called. Or what if you don’t know what’s going to get passed in? Some randomness or external method etc. 

Hope this helps

**Comment by afawcett** [comment]
@comic96 any views on this?

File: src/classes/fflib_Mocks.cls [hunk 1]
```diff
@@ -26,6 +26,16 @@ public class fflib_Mocks
...
 		}
 
+		public String get3(Integer index, String value, Boolean aBool)
+		{
+			return (String) mocks.mockNonVoidMethod(this, 'get3', new List<Object> {index, value, aBool});
+		}
+
+		public String get4(Integer index, String value, Boolean aBool, SObject aSObject)
+		{
+			return (String) mocks.mockNonVoidMethod(this, 'get3', new List<Object> {index, value, aBool, aSObject});
```
Comment (DanielTempleUK) [inline src/classes/fflib_Mocks.cls]: The String passed to mockNonVoidMethod should be `'get4'` for this method.

**Comment by cwilkinsonffdc** [comment]
Cheers @DanielTempleUK. Good spot. Just fixed that now.

**Comment by ImJohnMDaniel** [comment]
@afawcett and @comic96 

There seems to be some interesting features here.

Any thoughts about when the PR can be reviewed and merged in??

Cheers!

cc: @cwilkinsonffdc

**Comment by comic96** [comment]
Hi @cwilkinsonffdc as discussed, closing this PR since the implementation style is not in keeping with the overall design goal of ApexMocks being as close to Mockito as possible.

**Pull request closed without merging**